### ISSUE TITLE
Not Working on Ubuntu 16.04

### ISSUE THREAD

**Issue by Mattzi** [timschumi/gmod-chttp#8]
Hey, so I cant get it to work on my machine thats running Ubuntu 16.04. Regardless of which dll I use(normal 32bit and 32bit debian) it tells me: "Can't load module library". Can anyone help me out here?

**Comment by CoolGoga** [comment]
Hi! Look at the similar issue which may help you:

#2

**Comment by Mattzi** [comment]
@CoolGoga yeah I saw that but Im not able to install anything and just using the dll files from the releases

**Comment by timschumi** [comment]
@Mattzi Since you are using an older version of Ubuntu, please try using the DLLs suffixed with -deb9 (remember to rename them before use).

If that still doesn't help, please provide `ldd` output of both libraries (as outlined in issue #2)

**Comment by Mattzi** [comment]
@timschumi I tried those. I‘ll post that ldd stuff here later today

**Comment by Mattzi** [comment]
@timschumi everythings there except the libcurl thingy. But because I am using Docker I'm unable to install that and have just thrown the dll into my main folder. How can I get this working?

**Comment by Mattzi** [comment]
I now even uploaded the libcurl.so.4 file into the main folder next to my srcds and it still does not find it

**Comment by timschumi** [comment]
Keep in mind that the lib has to be the correct architecture (probably 32-bit by default). Also, you may have to add the dependencies of libcurl (for example openssl, in the rare case where it isn't installed), you can do that by repeatedly checking `ldd`.

Also, you will get more debug output if you add `LD_DEBUG=libs` as an environment variable (if you are able to do that). It will end up on the console, but it might get a bit spammed, so make sure that you are loading CHTTP after that has decreased.

If all that doesn't help, could you share which Docker image/Dockerfile you are using? Then I could try and set it up locally.

**Comment by Mattzi** [comment]
@timschumi so I downloaded 32bit Libcurl.dll from this repository aswell as uploaded the 32bit libcurl.so.4. How do you mean repeatedly checking ldd?

If nothing works i'll try that.

Probably not, It's the docker File of a Server Host thats run by a friend of mine

**Comment by timschumi** [comment]
> so I downloaded 32bit Libcurl.dll from this repository aswell as uploaded the 32bit libcurl.so.4.

You can't use Windows DLLs. The libcurl/libcrypto/libssl DLLs linked on the releases page are strictly Windows-only. The Linux libraries being suffixed with .dll is bullshittery on GMod's side.
Also, libraries might have dependencies on specific versions of glibc. Make sure that you are getting those libraries from the Ubuntu version you are trying to run this on (in your case, the 32-bit libraries from Ubuntu 16.04).

> How do you mean repeatedly checking ldd?

After adding libraries, run the ldd command you previously used and check if any library does now show up (or if there are new dependencies that have to be satisfied). Repeat that when making changes to see whether you are making progress.

>Probably not, It's the docker File of a Server Host thats run by a friend of mine

Taking a second look at the ldd output, that library folder structure doesn't look like Ubuntu's. Are you sure that the system inside the container is Ubuntu 16.04?

PS: I might have something in store that will help you with your problem. Basically, it's a CHTTP DLL that already has libcurl and its dependencies "built-in". It doesn't work for Windows yet and I haven't tested it extensively, so it's still unreleased. Let me know if you want to try that.

**Comment by Mattzi** [comment]
Oh well, I thought the dlls are for linux. I did get this libcurl.so.4 on Ubuntu 16.04 so that would be right.

I tried running ldd after adding the libcurl.so.4 but that did not change anything.

I can just tell you what my friend tells me and he said its Ubuntu 16.04 because he had bad experience when using the ubuntu 18 lts for gmod and the 20 lts is too new in his opinion.

That thing sounds like I would try it because we are still working on the server and its not that bad if any bugs happen. Hit me up if I can try that

**Comment by timschumi** [comment]
> Oh well, I thought the dlls are for linux. I did get this libcurl.so.4 on Ubuntu 16.04 so that would be right.
> 
> I tried running ldd after adding the libcurl.so.4 but that did not change anything.

As in "still says not found"? Is the working directory set correctly (i.e. the folder in which srcds_run is located)? I don't remember the specifics, but I think it only loads libs from the current working directory. Ask your friend if he is cd-ing into the server's root directory before starting it.

> I can just tell you what my friend tells me and he said its Ubuntu 16.04 because he had bad experience when using the ubuntu 18 lts for gmod and the 20 lts is too new in his opinion.

To be honest, at some point it would be easier to just preinstall 32-bit libcurl in the Docker image.

> That thing sounds like I would try it because we are still working on the server and its not that bad if any bugs happen. Hit me up if I can try that.

Then I'll get a file ready, make sure that it works in 16.04 (could you at least tell which base image you are using), and send it to you/upload it here.

**Comment by Mattzi** [comment]
@timschumi Yeah it still says not found.

I'll ask my friend the other two things and reply here

**Comment by Mattzi** [comment]
@timschumi he uses the standard ubuntu:16.04 Image with amd64 architecture

**Comment by Mattzi** [comment]
How long do you think will it take you to get that file ready? @timschumi

**Comment by timschumi** [comment]
> How long do you think will it take you to get that file ready? @timschumi

I had to make some more changes, since I forgot to include more libraries (which are usually installed by default, but not always). I hope that I can upload the file in the evening (~4 hours from now).

**Comment by timschumi** [comment]
Files are ready, they seem to work on the default Ubuntu 16.04 docker container without anything installed (except steamcmd, which automatically pulls in a few i386 dependencies).
Since those are now compiled on Debian 9 by default for compatibility reasons, there aren't any seperate -deb9 libraries anymore.

[static_gmsv_chttp.zip](https://github.com/timschumi/gmod-chttp/files/4587520/gmsv_chttp.zip)

**Comment by Mattzi** [comment]
Cant test much but at least the Server is loading chttp now

**Comment by timschumi** [comment]
Since I got static linking working on Windows, I now officially released a new revision. Hopefully, this will make installation issues like this obsolete.

**Comment by FlorianLeChat** [comment]
I still have the same problem as Mattzi despite having installed the latest version (release on May 8). But, I'm currently running on CentOS 7.

**Comment by timschumi** [comment]
CentOS 7 is years old at this point. I don't know what their policy with base library upgrades on minor versions is, but in the worst case, it might be 3 years older (or more) than the Debian 9 system that I'm building this on.

The problem Mattzi had (and that I took care of in the latest release) was the dependency on libcurl/OpenSSL. What you are running into here is probably the standard library (`libc`, etc.) having incompatible symbols (or the linker does not load those for any reason).

I can try building the libraries on Debian 8 for testing (which on its first release is still a year newer than the oldest CentOS 7 release, but maybe this is just enough). In the meantime, could you obtain the `ldd` output for the library, as outlined at the beginning of this ticket and in #2?

**Comment by FlorianLeChat** [comment]
Thank you for your response ! This is my ldd ouput :
```
/home/gmodserver/serverfiles/garrysmod/lua/bin/gmsv_chttp_linux.dll: /lib/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by /home/gmodserver/serverfiles/garrysmod/lua/bin/gmsv_chttp_linux.dll)
/home/gmodserver/serverfiles/garrysmod/lua/bin/gmsv_chttp_linux.dll: /lib/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by /home/gmodserver/serverfiles/garrysmod/lua/bin/gmsv_chttp_linux.dll)
linux-gate.so.1 =>  (0xf77c5000)
libstdc++.so.6 => /lib/libstdc++.so.6 (0xf7312000)
libm.so.6 => /lib/libm.so.6 (0xf72d0000)
libgcc_s.so.1 => /lib/libgcc_s.so.1 (0xf72b5000)
libc.so.6 => /lib/libc.so.6 (0xf70ea000)
/lib/ld-linux.so.2 (0xf77c6000)
```
Do you think the problem would be solved by installing CentOS 8 ?

**Comment by timschumi** [comment]
Yeah, your C++ standard library is too old.

Installing CentOS 8 would almost certainly resolve your problems, since it was released in 2019.
Although you should probably hold off on that, I first want to try if Debian 8 libraries are compatible (CentOS 7 is still widely used as an OS for servers, so I want to provide a different solution than "install CentOS 8").

**Comment by timschumi** [comment]
@FlorianLeChat I made a [new set of testing libraries](https://github.com/timschumi/gmod-chttp/files/4761809/gmsv_chttp_linux.zip), compiled on Debian 8. However, from what I've been able to determine, CentOS 7 still ships a version of glibc that is one version too old to run those libraries.

Also, in case you want to continue using CentOS 7 and those libraries didn't resolve the issue, please open a new issue. At this point, this discussion isn't relevant to this specific ticket anymore.

**Comment by FlorianLeChat** [comment]
Unfortunately those libraries didn't solve my problem. I'll probably switch to CentOS 8 or a more recent distribution in a few days, so that won't be a problem anymore. Thanks for your help anyway!
